File: frontend/src/commonMain/kotlin/com/farm/frontend/api/FarmApiClient.kt
// farm/frontend/src/commonMain/kotlin/com/farm/frontend/api/FarmApiClient.kt
package com.farm.frontend.api

import com.farm.common.Asset
import com.farm.common.FileDetail
import com.farm.common.UpdateAssetRequest
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.request.forms.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.Json

object FarmApiClient {
    private val client = HttpClient {
        install(ContentNegotiation) {
            json(Json {
                prettyPrint = true
                isLenient = true
                ignoreUnknownKeys = true
            })
        }
    }

    private const val BASE_URL = "/api" // Ktor backend routes start with /api

    suspend fun getAssets(): List<Asset> {
        return client.get("$BASE_URL/assets").body()
    }

    suspend fun getAssetDetails(assetId: Int): Asset? {
        return client.get("$BASE_URL/assets/$assetId").body()
    }

    suspend fun searchAssets(query: String): List<Asset> {
        return client.get("$BASE_URL/assets/search") {
            parameter("query", query)
        }.body()
    }

    suspend fun uploadAsset(
        assetName: String,
        link: String?,
        storeName: String?,
        authorName: String?,
        licenseName: String?,
        tags: String?,
        projects: String?,
        fileBytes: ByteArray,
        fileName: String,
        fileType: String
    ): String {
        return client.post("$BASE_URL/assets/upload") {
            setBody(MultiPartFormDataContent(
                formData {
                    append("asset_name", assetName)
                    link?.let { append("link", it) }
                    storeName?.let { append("store_name", it) }
                    authorName?.let { append("author_name", it) }
                    licenseName?.let { append("license_name", it) }
                    tags?.let { append("tags", it) }
                    projects?.let { append("projects", it) }
                    append("file", fileBytes, Headers.build {
                        append(HttpHeaders.ContentType, fileType)
                        append(HttpHeaders.ContentDisposition, "filename=\"$fileName\"")
                    })
                }
            ))
        }.bodyAsText()
    }

    suspend fun updateAsset(assetId: Int, request: UpdateAssetRequest): String {
        return client.put("$BASE_URL/assets/$assetId") {
            contentType(ContentType.Application.Json)
            setBody(request)
        }.bodyAsText()
    }
}

File: frontend-react/src/api/farmApiClient.ts
// farm/frontend-react/src/api/farmApiClient.ts
import { Asset, UpdateAssetRequest } from '../types/models';

const BASE_URL = '/api'; // Ktor backend routes start with /api

export const FarmApiClient = {
  /**
   * Fetches all assets from the backend.
   * @returns A promise that resolves to a list of Asset objects.
   */
  getAssets: async (): Promise<Asset[]> => {
    const response = await fetch(`${BASE_URL}/assets`);
    if (!response.ok) {
      throw new Error(`Failed to fetch assets: ${response.statusText}`);
    }
    return response.json();
  },

  /**
   * Fetches detailed information for a specific asset by its ID.
   * @param assetId The ID of the asset to fetch.
   * @returns A promise that resolves to an Asset object or null if not found.
   */
  getAssetDetails: async (assetId: number): Promise<Asset | null> => {
    const response = await fetch(`${BASE_URL}/assets/${assetId}`);
    if (response.status === 404) {
      return null;
    }
    if (!response.ok) {
      throw new Error(`Failed to fetch asset details: ${response.statusText}`);
    }
    return response.json();
  },

  /**
   * Searches for assets based on a query string.
   * @param query The search query string.
   * @returns A promise that resolves to a list of Asset objects.
   */
  searchAssets: async (query: string): Promise<Asset[]> => {
    const response = await fetch(`${BASE_URL}/assets/search?query=${encodeURIComponent(query)}`);
    if (!response.ok) {
      throw new Error(`Failed to search assets: ${response.statusText}`);
    }
    return response.json();
  },

  /**
   * Uploads a new asset including its file and metadata.
   * @param assetName The name of the asset.
   * @param link Optional link to the asset's source.
   * @param storeName Optional name of the store.
   * @param authorName Optional name of the author.
   * @param licenseName Optional name of the license.
   * @param tags Optional comma-separated tags.
   * @param projects Optional comma-separated projects.
   * @param file The File object to upload.
   * @returns A promise that resolves to a success message string.
   */
  uploadAsset: async (
    assetName: string,
    link: string | null,
    storeName: string | null,
    authorName: string | null,
    licenseName: string | null,
    tags: string | null,
    projects: string | null,
    file: File,
  ): Promise<string> => {
    const formData = new FormData();
    formData.append('asset_name', assetName);
    if (link) formData.append('link', link);
    if (storeName) formData.append('store_name', storeName);
    if (authorName) formData.append('author_name', authorName);
    if (licenseName) formData.append('license_name', licenseName);
    if (tags) formData.append('tags', tags);
    if (projects) formData.append('projects', projects);
    formData.append('file', file); // Append the actual file

    const response = await fetch(`${BASE_URL}/assets/upload`, {
      method: 'POST',
      body: formData, // FormData automatically sets 'Content-Type: multipart/form-data'
    });

    const textResponse = await response.text(); // Read as text, as backend returns string message
    if (!response.ok) {
      throw new Error(`Upload failed: ${textResponse}`);
    }
    return textResponse;
  },

  /**
   * Updates an existing asset's metadata.
   * @param assetId The ID of the asset to update.
   * @param request The UpdateAssetRequest object containing the new metadata.
   * @returns A promise that resolves to a success message string.
   */
  updateAsset: async (assetId: number, request: UpdateAssetRequest): Promise<string> => {
    const response = await fetch(`${BASE_URL}/assets/${assetId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    const textResponse = await response.text();
    if (!response.ok) {
      throw new Error(`Update failed: ${textResponse}`);
    }
    return textResponse;
  },
};